File: java/leet_code/165.java
/**
 * This class provides a solution for comparing version numbers.
 * The version numbers are represented as strings, and the comparison is done by comparing each segment of the version number.
 * The comparison stops as soon as a difference is found.
 * If no difference is found after comparing all segments, the versions are considered equal.
 */
class Solution {

    /**
     * Compares two version numbers.
     *
     * @param version1 the first version number as a string. It is expected to be a non-null string where segments of the version number are separated by '.'.
     * @param version2 the second version number as a string. It is expected to be a non-null string where segments of the version number are separated by '.'.
     *
     * @return an integer indicating the result of the comparison. It returns -1 if version1 is less than version2, 1 if version1 is greater than version2, and 0 if both versions are equal.
     */
    public int compareVersion(String version1, String version2) {
        // Initialize pointers for both versions
        int i = 0, j = 0;

        // Continue comparison as long as there are segments in either version
        while (i < version1.length() || j < version2.length()) {
            // Initialize segment values for this round of comparison
            int num1 = 0, num2 = 0;

            // Construct the segment from version1
            while (i < version1.length() && version1.charAt(i) != '.') {
                num1 = num1 * 10 + (version1.charAt(i) - '0');
                i++;
            }

            // Construct the segment from version2
            while (j < version2.length() && version2.charAt(j) != '.') {
                num2 = num2 * 10 + (version2.charAt(j) - '0');
                j++;
            }

            // Compare the segments
            if (num1 < num2) {
                return -1;
            }
            if (num1 > num2) {
                return 1;
            }

            // Move to the next segment
            i++;
            j++;
        }

        // If all segments are equal, the versions are equal
        return 0;
    }
}

File: python/leet_code/165.py
"""leet_code/165.py

This module contains a solution for comparing two version numbers (version1 and version2).
The version numbers are in the format of 'x.y.z....' where x, y, z are non-negative integers.
The comparison is done by comparing each corresponding pair of integer values in the version numbers.

Classes:
    Solution: This class contains a method to compare two version numbers.
"""


class Solution:
    """A class used to compare two version numbers.

    Methods:
        compareVersion: Compares two version numbers and returns an integer indicating the result.
    """

    def compareVersion(self, version1: str, version2: str) -> int:
        """Compares two version numbers.

        Args:
            version1 (str): The first version number in the format 'x.y.z....'.
            version2 (str): The second version number in the format 'x.y.z....'.

        Returns:
            int: Returns 1 if version1 is greater than version2, -1 if version1 is less than version2,
                 and 0 if both versions are equal.

        This function does not handle errors explicitly. If the input is not in the expected format,
        Python's built-in error handling will raise an exception. For example, if the version numbers
        do not contain integers or if they are not separated by '.', a ValueError will be raised.
        """
        # Split the version numbers by '.'
        version1 = version1.split('.')
        version2 = version2.split('.')

        # Get the minimum length of the two version numbers
        ml = min(len(version1), len(version2))

        # Compare each corresponding pair of integers in the version numbers
        for i in range(ml):
            v1 = int(version1[i])
            v2 = int(version2[i])
            if v1 < v2:
                return -1
            elif v1 > v2:
                return 1

        # If the version numbers have different lengths, check the remaining integers in the longer version number
        version = version1 if len(version1) - ml > 0 else version2
        v = 1 if len(version1) - ml > 0 else 2
        for i in range(ml, len(version)):
            if int(version[i]) > 0:
                return 1 if v == 1 else -1

        # If all corresponding pairs of integers are equal, the version numbers are equal
        return 0
